src/representations/feature_manager_block_sparse.hh

```diff
@@ -264,6 +291,204 @@ namespace rascal {
     std::set<Key_t> unique_keys{};
   };
 
+  template <typename Precision_t>
+  using Dense_t = Eigen::Matrix<Precision_t, Eigen::Dynamic, Eigen::Dynamic>;
+
+  template <typename T>
+  std::list<T> intersection_of(const std::list<T> & a, const std::list<T> & b) {
+    std::list<T> rtn;
+    std::unordered_multiset<T, internal::Hash<T>> st;
+    std::for_each(a.begin(), a.end(), [&st](const T & k) { st.insert(k); });
+    std::for_each(b.begin(), b.end(), [&st, &rtn](const T & k) {
+      auto iter = st.find(k);
+      if (iter != st.end()) {
+        rtn.push_back(k);
+        st.erase(iter);
+      }
+    });
+    return rtn;
+  }
+
+  /**
+   * Matrix multiplication between two block sparse feature matrices
+   */
+  template <typename Precision_t>
+  decltype(auto) dot(FeatureManagerBlockSparse<Precision_t> & X1,
+                     FeatureManagerBlockSparse<Precision_t> & X2) {
+    auto n1_centers{X1.sample_size()};
+    auto n2_centers{X2.sample_size()};
+    Dense_t<Precision_t> mat =
```

Comment:
Maybe I am getting the math wrong here (which is totally possible), therefore for clarification:
When two block sparse feature matrices from two different structures where not keys of no centers intersect, the matrix here stays zero, like initialized.
And that is what is actually wanted, right?

Reply:
yes in this case. one could introduce some coupling but I would rather do this kind of operation before hand

Reply:
Thanks for the small benchmark these small good practice add up in the end.

    feature_manager_block_sparse: I think we will need a schematic for the layout
    of the block sparse layout. The description in the beginning of the file is
    helpful, but a drawing would make it much more visual. I get the principle idea, but I am lost on the details. I tried to comment on the parts, which do the actual math.
1. since this will disappear soon I would rather not spend too much time on the details here. If you have questions I can always elaborate though.
2. there are not tests since this is temporary. Michele wants the functionality available now but the overall implementation will have to change quite a bit latter so the tests seems a bit too much to me. Of course I ran some comparisons and checks which allowed me to track down a bug so the computation works.